packages/schematics/angular/utility/workspace-models.ts

```diff
@@ -0,0 +1,150 @@
+    Protractor = '@angular-devkit/build-angular:protractor',
+}
+
+export interface FileReplacements {
+    replace: string;
+    with: string;
+}
+
+export interface BrowserBuilderBaseOptions {
+    main: string;
+    tsConfig: string;
+    fileReplacements?: FileReplacements[];
+    outputPath?: string;
+    index?: string;
+    polyfills: string;
+    assets?: object[];
+    styles?: string[];
+    scripts?: string[];
+    sourceMap?: boolean;
+}
+
+export interface BrowserBuilderOptions extends BrowserBuilderBaseOptions {
+    serviceWorker?: boolean;
+    optimization?: boolean;
+    outputHashing?: 'all';
+    extractCss?: boolean;
+    namedChunks?: boolean;
+    aot?: boolean;
+    extractLicenses?: boolean;
+    vendorChunk?: boolean;
+    buildOptimizer?: boolean;
+}
+
+export interface ServeBuilderOptions {
+    browserTarget: string;
+}
+export interface LibraryBuilderOptions {
+    tsConfig: string;
+    project: string;
+}
+
+export interface ServerBuilderOptions {
+    outputPath: string;
+    tsConfig: string;
+    main: string;
+    fileReplacements?: FileReplacements[];
+}
+
+export interface AppShellBuilderOptions {
+    browserTarget: string;
+    serverTarget: string;
+    route: string;
+}
+
+export interface TestBuilderOptions extends Partial<BrowserBuilderBaseOptions> {
+    karmaConfig: string;
+}
+
+export interface LintBuilderOptions {
+    tsConfig: string[] | string;
+    exclude?: string[];
+}
+
+export interface ExtractI18nOptions {
+    browserTarget: string;
+}
+
+export interface E2EOptions {
+    protractorConfig: string;
+    devServerTarget: string;
+}
+
+export interface BuilderTarget<TBuilder extends Builders, TOptions> {
+    builder: TBuilder;
+    options: TOptions;
+    configurations?: {
+        production: Partial<TOptions>;
+        [key: string]: Partial<TOptions>;
+    };
+}
+
+export type LibraryBuilderTarget = BuilderTarget<Builders.NgPackagr, LibraryBuilderOptions>;
+export type BrowserBuilderTarget = BuilderTarget<Builders.Browser, BrowserBuilderOptions>;
+export type ServerBuilderTarget = BuilderTarget<Builders.Server, ServerBuilderOptions>;
+export type AppShellBuilderTarget = BuilderTarget<Builders.AppShell, AppShellBuilderOptions>;
+export type LintBuilderTarget = BuilderTarget<Builders.TsLint, LintBuilderOptions>;
+export type TestBuilderTarget = BuilderTarget<Builders.Karma, TestBuilderOptions>;
+export type ServeBuilderTarget = BuilderTarget<Builders.DevServer, ServeBuilderOptions>;
+export type ExtractI18nBuilderTarget = BuilderTarget<Builders.ExtractI18n, ExtractI18nOptions>;
+export type E2EBuilderTarget = BuilderTarget<Builders.Protractor, E2EOptions>;
+
+export interface WorkspaceSchema extends experimental.workspace.WorkspaceSchema {
+    projects: {
+        [key: string]: WorkspaceProject<ProjectType.Application | ProjectType.Library>;
+    };
+}
+
+// tslint:disable-next-line:no-any
+export interface WorkspaceProject<TProjectType extends ProjectType = any>
```

Comment:
Could we do without the `any` here? The main usecase seems to be `ProjectType.Application` so that seems a sensible default.

Reply:
The thing is that having it as a default would break a lot of usages, and would require a bit of boilerplate for usages. Example;
```ts
const project = getProject<ProjectType.Application>("foo");
```
Because it will have a cascading effect with such errors;
```
'WorkspaceProject<ProjectType>' is not assignable to type 'WorkspaceProject<ProjectType.Application>'.
```

Alternativly if you prefer we can do two methods 
```ts
const project1 = getLibraryProject("foo");
const project2 = getApplicationProject("foo");
```

Reply:
With the diff below I think the changes are complete. 
```
$ git diff
diff --git a/packages/schematics/angular/utility/config.ts b/packages/schematics/angular/utility/config.ts
index 5033f77c5..a1e0018b1 100644
--- a/packages/schematics/angular/utility/config.ts
+++ b/packages/schematics/angular/utility/config.ts
@@ -7,7 +7,7 @@
  */
 import { JsonParseMode, parseJson } from '@angular-devkit/core';
 import { Rule, SchematicContext, SchematicsException, Tree } from '@angular-devkit/schematics';
-import { WorkspaceProject, WorkspaceSchema } from './workspace-models';
+import { ProjectType, WorkspaceProject, WorkspaceSchema } from './workspace-models';


 // The interfaces below are generated from the Angular CLI configuration schema
@@ -497,10 +497,10 @@ export function getWorkspace(host: Tree): WorkspaceSchema {
   return parseJson(content, JsonParseMode.Loose) as {} as WorkspaceSchema;
 }

-export function addProjectToWorkspace(
+export function addProjectToWorkspace<TProjectType extends ProjectType = ProjectType.Application>(
   workspace: WorkspaceSchema,
   name: string,
-  project: WorkspaceProject,
+  project: WorkspaceProject<TProjectType>,
 ): Rule {
   return (host: Tree, context: SchematicContext) => {

diff --git a/packages/schematics/angular/utility/project.ts b/packages/schematics/angular/utility/project.ts
index caf3c1c76..13df722f0 100644
--- a/packages/schematics/angular/utility/project.ts
+++ b/packages/schematics/angular/utility/project.ts
@@ -24,15 +24,15 @@ export function buildDefaultPath(project: WorkspaceProject): string {
   return `${root}${projectDirName}`;
 }

-export function getProject(
+export function getProject<TProjectType extends ProjectType = ProjectType.Application>(
   workspaceOrHost: WorkspaceSchema | Tree,
   projectName: string,
-): WorkspaceProject {
+): WorkspaceProject<TProjectType> {
   const workspace = isWorkspaceSchema(workspaceOrHost)
     ? workspaceOrHost
     : getWorkspace(workspaceOrHost);

-  return workspace.projects[projectName];
+  return workspace.projects[projectName] as WorkspaceProject<TProjectType>;
 }

 export function isWorkspaceSchema(workspace: unknown): workspace is WorkspaceSchema {
diff --git a/packages/schematics/angular/utility/workspace-models.ts b/packages/schematics/angular/utility/workspace-models.ts
index 6eb5c9a4f..b262e26d1 100644
--- a/packages/schematics/angular/utility/workspace-models.ts
+++ b/packages/schematics/angular/utility/workspace-models.ts
@@ -119,8 +119,7 @@ export interface WorkspaceSchema extends experimental.workspace.WorkspaceSchema
     };
 }

-// tslint:disable-next-line:no-any
-export interface WorkspaceProject<TProjectType extends ProjectType = any>
+export interface WorkspaceProject<TProjectType extends ProjectType = ProjectType.Application>
     extends experimental.workspace.WorkspaceProject {
     /**
     * Project type.
```

In `packages/schematics/angular/library/index.ts`, line 172, the type is then inferred as `ProjectType.Library` with no need to specify it:
![image](https://user-images.githubusercontent.com/4172079/45357781-4a04a700-b5bf-11e8-91d0-641c66a2505e.png)

Reply:
@filipesilva thanks so much for this. Feeling silly, that I didn't think of it before.

Reply:
It's a bit more work but in the past it has paid dividends. Usually every `any` we manage to get rid of prevents regressions down the line.